content/dharma/abstract-data-structures.md

```diff
@@ -0,0 +1,118 @@
+
+In contrast to the data structures which are specific and detailed implementations that deal with how the data structure does its job, abstract data type focuses on _what it does_ and not how it does its job. In short, the ADT defines what that particular data construct must do and the data structure is the concrete implementation of that construct.
+
+An analogy to explain abstract data type in terms of web development would be CRUD APIs. The user of any CRUD API has to simply know what request method(GET, POST, etc.) should they send and if they followed the rules of the API, the API server would send data back. The user didn't have to worry about the internal workings of the API server. The user simply had to know the rules of interactions and behavior of CRUD API. In this case, the CRUD API is functioning as an abstract data type _from the perspective of the user_.
+
+There are no specific rules which force the implementation of particular methods and operations in a particular ADT. This is decided based on the requirements in a use-case scenario and therefore, ultimately a purely design decision.
+
+## Why use ADTs?
+
+There are 3 general advantages of using ADTs, listed as follows:
+
+### Encapsulation
+
+ADT will provide certain methods and properties and knowledge of these methods and properties is all the user will need to successfully operate with the ADT
+
+### Compartmentalization
+
+The code that is using the ADT will not have to be changed even if the internal workings of the ADT have been changed. The change in the ADT is isolated and compartmentalized.
+
+### Adaptability
+
+Real world programs continue to evolve with ever changing requirements or new constraints. Differently implemented ADTs, with all the same properties and methods, can thus be used interchangebly. For example, if there was linked list created using lists/arrays that contained name of the patients in a hospital but later they decide to include all the information about the patient in the linked list, then a linked list implemented using class based nodes with all the necessary fields would serve as a much better replacement as compared to the linked list that is simply using arrays/lists.
+
+## General Operations supported by ADTs
+
+ADTs support the follow operations:
+
+1. **Traversing**: Traversing allows each data in the ADT to be accessed once for processing
+2. **Searching**: Searching allows the user to look for specific data in the ADT
+3. **Inserting**: Inserting allows the user to insert data at a particular index/space in the ADT
+4. **Deleting**: Deleting allows the user to either delete a particular piece of data or delete data at a particular location
+5. **Sorting**: Sorting operation would allow the data to be ordered in ascending or descending order, depending on the preference.
```

Comment:
```suggestion
5. **Sorting**: Sorting operation would allow the elements to be ordered in ascending or descending order, depending on the preference.
```